File: src/main/java/shane/leetcode/problems/easy/Q349IntersectionOfTwoArrays.java
package shane.leetcode.problems.easy;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Set;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;

public class Q349IntersectionOfTwoArrays {

    @Test
    public void test() {
        assertThat(intersection(new int[]{1, 2, 2, 1}, new int[]{2, 2})).containsExactlyInAnyOrder(2);
        assertThat(intersection(new int[]{4, 9, 5}, new int[]{9, 4, 9, 8, 4})).containsExactlyInAnyOrder(9, 4);
    }

    public int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> set1 = Arrays.stream(nums1).boxed().collect(Collectors.toSet());
        Set<Integer> set2 = Arrays.stream(nums2).boxed().collect(Collectors.toSet());
        set1.retainAll(set2);
        return set1.stream().mapToInt(Integer::valueOf).toArray();
    }

}

File: src/main/kotlin/shane/leetcode/problems/easy/Q349IntersectionOfTwoArrays.kt
package shane.leetcode.problems.easy

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test

/**
 * Runtime: 455 ms, faster than 5.13% of Kotlin online submissions for Intersection of Two Arrays.
Memory Usage: 42.1 MB, less than 20.51% of Kotlin online submissions for Intersection of Two Arrays.
 */
class Q349IntersectionOfTwoArrays {

    @Test
    fun test() {
        Assertions.assertThat(intersection(intArrayOf(1, 2, 2, 1), intArrayOf(2, 2))).containsExactlyInAnyOrder(2)
        Assertions.assertThat(intersection(intArrayOf(4, 9, 5), intArrayOf(9, 4, 9, 8, 4)))
            .containsExactlyInAnyOrder(4, 9)
    }

    fun intersection(nums1: IntArray, nums2: IntArray): IntArray {
        return nums1.toSet().intersect(nums2.toSet()).toIntArray()
    }
}
